src/pages/experiments/[experimentId]/data-exploration/components/heatmap/HeatmapPlot.jsx

```diff
@@ -61,36 +61,63 @@ const HeatmapPlot = (props) => {
   }, [loadingGenes, hidden]);
 
   const createVegaData = (selected, expression) => {
-    const data = { cellOrder: [], geneOrder: [], heatmapData: [] };
-
-    data.geneOrder = selected;
-
-    data.cellOrder = [];
+    const data = { cellOrder: [], geneOrder: selected, heatmapData: [] };
 
     // Get all hidden cells
     const hiddenCellIds = union(Array.from(hidden), properties);
 
+    // Downsample cells.
+    let totalCellsToShow = 1000;
+    const groupByCluster = {};
+
     const louvainClusters = hierarchy.filter((clusters) => clusters.key === 'louvain');
     if (louvainClusters.length > 0) {
+      let totalCellsFetched = 0;
+
       const clusterKeys = louvainClusters[0].children;
 
       clusterKeys.forEach(({ key }) => {
         const cellsToShow = Array.from(
           properties[key].cellIds,
         ).filter((id) => !hiddenCellIds.has(id));
 
-        data.cellOrder.push(...cellsToShow);
+        groupByCluster[key] = cellsToShow;
+        totalCellsFetched += cellsToShow.length;
+      });
+
+      // We show the number of total cells fetched or the default value,
+      // whichever is smaller.
+      totalCellsToShow = Math.min(totalCellsToShow, totalCellsFetched);
+
+      // Create a sample of cells to display.
+      clusterKeys.forEach(({ key }) => {
+        const cells = groupByCluster[key];
+
+        if (!cells) {
+          return;
+        }
+
+        // Create a sample size proportional to the number of cells to show
+        // as well as the number of cells in the cluster.
+        const sampleSize = Math.floor(
+          (cells.length / totalCellsFetched) * totalCellsToShow,
```

Comment:
is there any chance that `totalCellsFetched` can be 0? Because if there is, then this line will blow up

Reply:
I guess it is theoretically possible. I will add a guard for it.